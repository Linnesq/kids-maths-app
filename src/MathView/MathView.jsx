import React from 'react';
import PropTypes from 'prop-types';

import DisplayBox from '../DisplayBox';
import InputBox from '../InputBox';
import SubmitButton from '../SubmitButton'
import TaskOptionsPanel from '../TaskOptionsPanel';

import './MathView.scss';

const results = {
    UNKNOWN: 0,
    INCORRECT: 1,
    CORRECT: 2,
};

class MathView extends React.Component {
    constructor(props) {
        super(props);
        const { controls, taskFunctions } = props.taskData
        const { userFills } = props.taskData.display;
        const { range, factor } = controls[0];

        this.state = {
            result: results.UNKNOWN,
            value: "",
            displayText: controls[0].displayText, // TODO, this is acting as an Unique ID
            taskType: controls[0].taskType,
            task: taskFunctions[controls[0].taskType].apply(this, [userFills, range, factor]),
            range,
            factor,
        };
    }

    componentDidMount() {
        // eslint-disable-next-line no-console
        console.debug('component mounted');
    }

    componentWillUnmount() {
        // eslint-disable-next-line no-console
        console.debug('component unmounting');
    }

    onSubmitHandler = () => {
        if (this.state.result === results.CORRECT) {
            this.onNewTask();
        } else {
            const isCorrect = this.state.value === this.state.task[this.state.task.userFills];
            this.setState({
                result: isCorrect ? results.CORRECT : results.INCORRECT,
                value: isCorrect ? this.state.value : "" });
        }
    }

    getTaskCreator = () => {
        return this.props.taskData.taskFunctions[this.state.taskType];
    }

    getTaskCreatorArgs = () => [this.props.taskData.display.userFills, this.state.range, this.state.factor];

    onNewTask = () => {
        this.setState({
            task: this.getTaskCreator().apply(this, this.getTaskCreatorArgs()),
            value: "",
            result: results.UNKNOWN,
        });
    }

    onTaskTypeChange = (displayText) => {
        const newTask = this.props.taskData.controls.filter(item => item.displayText === displayText)[0];
        const newFunction = this.props.taskData.taskFunctions[newTask.taskType];
        const { range, factor } = newTask;
        const args = [this.props.taskData.display.userFills, range, factor];
        
        this.setState({
            taskType: newTask.taskType,
            displayText: newTask.displayText,
            value: "",
            result: null,
            task: newFunction.apply(this, args),
            range,
            factor
        });
    }

    onInputValueChange = (newValue) => {
        this.setState({
            value: newValue,
        });
    }

    render(){
        const inputBox = <InputBox onValueChangeHandler={this.onInputValueChange} intValue={this.state.value} onSubmit={this.onSubmitHandler} />;
        const { controls } = this.props.taskData;
        const { left, right, answer } = this.state.task;
        let resultsText = "Do your best!";
        if (this.state.result === results.CORRECT){
            resultsText = "Correct, well done!";
        } else if (this.state.result === results.INCORRECT) {
            resultsText = "Unlucky, keep trying!";
        }

        const submitButtonText = this.state.result === results.CORRECT ? "Next" : "Submit";

        return (
            <>
                <div className="MathView">
                    {this.state.task.userFills === "left" ? inputBox : <DisplayBox value={left} />}
                    <span className="MathView-sign">{this.state.task.operator}</span>
                    {this.state.task.userFills === "right" ? inputBox : <DisplayBox value={right}/>}
                    <span className="MathView-sign">=</span>
                    {this.state.task.userFills === "answer" ? inputBox : <DisplayBox value={answer} />}
                </div>
                    <SubmitButton onClickHandler={this.onSubmitHandler} buttonText={submitButtonText} />
                <div>
                    <h3>{resultsText}</h3>
                    <button onClick={this.onNewTask}>New Math Question</button>
                </div>
                <TaskOptionsPanel
                    onClickHandler={this.onTaskTypeChange}
                    displayText={this.state.displayText}
                    controlInfo={controls} />
            </>
        )
    }
}

MathView.propTypes = {
    taskData: PropTypes.shape({
        controls: PropTypes.arrayOf(PropTypes.shape({
            taskType: PropTypes.string.isRequired,
            displayText: PropTypes.string.isRequired,
            range: PropTypes.number,
            factor: PropTypes.number,
    })),
        taskFunctions: PropTypes.objectOf(PropTypes.func),
        display: PropTypes.shape({
            userFills: PropTypes.oneOf(['left', 'right', 'random']).isRequired,
        }),
    }),
};

export default MathView;