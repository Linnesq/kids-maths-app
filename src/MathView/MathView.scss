@import "../variables.scss";

.MathView {
  display: inline-flex;
  width: 80%;
  margin-top: 2rem;
}

.MathView-sign {
  width: 20%;
  font-size: $textRegular;
}
