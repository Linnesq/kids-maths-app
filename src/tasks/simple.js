const {
    getSimpleAddition,
    getSimpleMultiplication,
    getSimpleDivision,
    getSimpleSubtraction,
    getTimesTable,
    getDivision,
} = require('../utils/math-tasks');

const SYMBOLS = {
    ADD: "+",
    SUB: "−",
    MULT: "×",
    DIV: "÷",
};

const USERFILLS = {
    LEFT: 'left',
    RIGHT: 'right',
    RANDOM: 'random',
};

const getSimpleTasks = () => {
    return {
    "description": "Multi-choice arithemetic",
    "controls": [
        { "displayText": SYMBOLS.ADD, "taskType": "addition" },
        { "displayText": SYMBOLS.MULT, "taskType": "multiplication" },
        { "displayText": SYMBOLS.DIV, "taskType": "division" },
        { "displayText": SYMBOLS.SUB, "taskType": "subtraction" },
    ],
    "taskFunctions": {
        "addition": getSimpleAddition,
        "multiplication": getSimpleMultiplication,
        "division": getSimpleDivision,
        "subtraction" : getSimpleSubtraction,
    },
    "display": {
        "userFills": USERFILLS.RANDOM,
    },
};
};

const getTimesTablesTask = () => {
    return {
        "description": "times tables",
        "controls": [
            { "displayText": "2X", "taskType" : "multiplication", "range": 12, "factor": 2 },
            { "displayText": "3X", "taskType" : "multiplication", "range": 12, "factor": 3 },
            { "displayText": "4X", "taskType" : "multiplication", "range": 12, "factor": 4 },
            { "displayText": "5X", "taskType" : "multiplication", "range": 12, "factor": 5 },
            { "displayText": "10X", "taskType" : "multiplication", "range": 12, "factor": 10 },
        ],
        "taskFunctions": {
            "multiplication": getTimesTable,
        },
        "display": {
            "userFills": USERFILLS.RANDOM,
        },
    };
};

const getDivisionTasks = () => {
    const division = 'division';

    return {
        "description": division,
        controls: [
            { displayText: "÷2", taskType: division, range: 12, factor: 2 },
            { displayText: "÷3", taskType: division, range: 12, factor: 3 },
            { displayText: "÷4", taskType: division, range: 12, factor: 4 },
            { displayText: "÷5", taskType: division, range: 12, factor: 5 },
        ],
        taskFunctions: {
            [division]: getDivision,

        },
        display: {
            userFills: USERFILLS.RANDOM,
        },
    };
};

export { getSimpleTasks, getTimesTablesTask, getDivisionTasks };