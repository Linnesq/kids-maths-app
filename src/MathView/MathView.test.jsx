import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import MathView from './MathView';
import { getDivisionTasks } from '../tasks/simple';

let mockMath;

const mockProps = () => {
    let mockProps = getDivisionTasks();
    mockProps.display.userFills = 'answer';
    return mockProps;
};

describe("MathView Component", () => {
    beforeEach(() => mockMath = jest.spyOn(global.Math, 'random').mockReturnValue(0.2));
    
    afterEach(() => mockMath.mockRestore());

    test('should initialise with maths problem', () => {
        const { container } = render(<MathView taskData={mockProps()} />);
        const [left, right] = container.getElementsByClassName("DisplayBox");

        expect(left.textContent).toEqual(expect.stringMatching('6'));
        expect(right.textContent).toEqual(expect.stringMatching('2'));
    });

    test('should accept input from keyboard and clicks', () => {
        const { container, getByText } = render(<MathView taskData={mockProps()} />);
        let submitButton = container.querySelector('.SubmitButton');
        expect(submitButton.textContent).toEqual("Submit");

        const inputBox = container.querySelector('.InputBox');
        fireEvent.change(inputBox, { target: { value: 3 } }); // input '3'
        fireEvent.keyUp(inputBox, { keyCode: 13 });  // press enter

        expect(submitButton.textContent).toEqual("Next");
        expect(getByText("Correct, well done!")).toBeInTheDocument();

        fireEvent.click(getByText("Next"));
        expect(getByText("Do your best!")).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    test("should not accept incorrect answers", () => {
        const { container, getByText } = render(<MathView taskData={mockProps()} />);

        const inputBox = container.querySelector('.InputBox');
        fireEvent.change(inputBox, { target: { value: 10 } }); // input '10'

        fireEvent.click(getByText("Submit"));
        expect(getByText("Unlucky, keep trying!")).toBeInTheDocument();
    });
});