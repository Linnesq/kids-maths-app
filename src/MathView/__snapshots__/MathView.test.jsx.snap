// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MathView Component should accept input from keyboard and clicks 1`] = `
<div>
  <div
    class="MathView"
  >
    <div
      class="DisplayBox"
    >
      <span>
        6
      </span>
    </div>
    <span
      class="MathView-sign"
    >
      ÷
    </span>
    <div
      class="DisplayBox"
    >
      <span>
        2
      </span>
    </div>
    <span
      class="MathView-sign"
    >
      =
    </span>
    <input
      class="InputBox"
      type="number"
      value=""
    />
  </div>
  <div
    class="SubmitButton"
  >
    <button
      data-testid="submitButton"
      type="submit"
    >
      Submit
    </button>
  </div>
  <div>
    <h3>
      Do your best!
    </h3>
  </div>
  <div>
    <button
      class="TaskOptionsPanel TaskOptionsPanel-selected"
      type="button"
    >
      ÷2
    </button>
    <button
      class="TaskOptionsPanel"
      type="button"
    >
      ÷3
    </button>
    <button
      class="TaskOptionsPanel"
      type="button"
    >
      ÷4
    </button>
    <button
      class="TaskOptionsPanel"
      type="button"
    >
      ÷5
    </button>
  </div>
</div>
`;
